Add support for component metadata supplier

melix opened the pull request:
This pull request adds support for a new concept: a component metadata supplier. This supplier is used whenever dynamic versions are in use, on Ivy repositories only for now.

Whenever a dynamic version is requested:

- we use the standard version lister to fetch the list of candidate versions for a specific module identifier
- for each version:
   - we call the supplier, which _can_ provide the "shallow" metadata for this component. In particular, its _status_ which is used for selection
   - if the status matches the expected version status, we stop there

This "rule" allows users to implement faster dynamic dependency resolution on Ivy repositories, by avoiding parsing Ivy descriptors for versions that will _not_ be included in the graph. In particular, for a module which is published often, here's what typically happens today, without the PR, when looking for `latest.release`:

- we perform a remote request to get the list of versions. Returns: `1.255`, `1.254`, `1.253`, ... `1.0`
- by descending order, we fetch the `ivy.xml` descriptor of `1.255`, parse it (completely, with full metadata information including dependencies) and find out its `status` is `integration`.
- goes on to the next version: `1.254`, and same again and again until we find the first version which status is `release`.
- stop and add the dependency metadata to the graph

Sometimes, we can benefit from caching of external resources, for versions which were already looked up before, but we're doing a lot of network requests, as well as a lot of parsing, just to get the `status` of a version.

With this PR, the process is optimized:

- we perform a remote request to get the list of versions. Returns: `1.255`, `1.254`, `1.253`, ... `1.0`
- by descending order, we call the supplier, so that it returns the `status` of the version. The supplier implementation is done by the user, who can avoid parsing a descriptor
- once the version with the appropriate status is found, we parse the descriptor, but only for this version, to get the full metadata and include it into the graph

Technically, it means that we can limit the number of requests done on the network:

- one for the list of versions
- one for each module, returning the status of all versions

Which makes it 2 requests for a single module, instead of  `1+O(nb of modules published since last checked)`.
Optionally, it is also possible to perform a _single request_ for all modules, which means 1 request per module to list the versions + 1 (cached) request for all modules in the graph. In other words, it would make only 1 request per module on average.

To implement this, a supplier can have access to a `RepositoryResourceAccessor`. Such an accessor needs to be _injected_ to the supplier, and provides a way to perform _cached requests_ on the repository, relative to its base URI.

A new supplier is created by Gradle for each module being queried. It is parallel safe, unless the user uses static fields in the implementation.

Here's an example of such a supplier:

```
class MP implements ComponentMetadataSupplier {
          
  final RepositoryResourceAccessor repositoryResourceAccessor
            
  @Inject
  MP(RepositoryResourceAccessor accessor) { repositoryResourceAccessor = accessor }
          
  void supply(ComponentMetadataSupplierDetails details) {
      def id = details.id
      println "Providing metadata for \$id"
      repositoryResourceAccessor.withResource("\${id.group}/\${id.module}/\${id.version}/status.txt") {
          details.result.status = new String(it.bytes)
      }
  }
}
```

This supplier will perform a request, based on the `group`, `module` and `version` of a module, which means, one request per module _version_. This version doesn't benefit much from network caching, because we fetch a metadata file (`status.txt`) per version. However, this version would be called for each version of the module, but a single network request would be done:

```
class MP implements ComponentMetadataSupplier {
          
  final RepositoryResourceAccessor repositoryResourceAccessor
            
  @Inject
  MP(RepositoryResourceAccessor accessor) { repositoryResourceAccessor = accessor }
          
  void supply(ComponentMetadataSupplierDetails details) {
      def id = details.id
      println "Providing metadata for \$id"
      repositoryResourceAccessor.withResource("\${id.group}/\${id.module}/status.txt") {
          // status.txt describes all versions of the module, not a single one
      }
  }
}
```

And of course, it is even possible to have one request for all modules, independently of their version:

```
class MP implements ComponentMetadataSupplier {
          
  final RepositoryResourceAccessor repositoryResourceAccessor
            
  @Inject
  MP(RepositoryResourceAccessor accessor) { repositoryResourceAccessor = accessor }
          
  void supply(ComponentMetadataSupplierDetails details) {
      def id = details.id
      println "Providing metadata for \$id"
      repositoryResourceAccessor.withResource("status.txt") {
          // status.txt describes all versions of the all modules
      }
  }
}
```

Usage of the rule is done in the repository description:

```
repositories {
    ivy {
        name 'repo'
        url 'https://....'
        metadataSupplier(MP) // use the custom metadata supplier
    }
}
```

Gradle will create a new instance of the supplier for _each module being queried_.
The cache provided by `RepositoryResourceAccessor` honors the semantics of `resolutionStrategy.cacheDynamicVersionsFor`.

melix commented:
Rebased onto `release`.

bigdaz commented:
The detailed description is great. Thanks.
Some initial feedback without looking at the actual implementation:
- There are probably use cases where the `status` of a module can be determined directly by inspecting the version, but in a way that's not easy to match via a version selector. These new rules would help in this case, without requiring _any_ network access to determine the status.
- Will we execute the `ComponentMetadataSupplier` if the actual component metadata has previously been downloaded and is in the cache? There will be cases where it would be inefficient to use the rule when we already have the full metadata available. (For one, the metadata is in a binary store and won't require parsing a cached resource).
- Is there a way to extract the default behaviour for dynamic versions (download the `ivy.xml` file and cache the metadata) as a built-in `ComponentMetadataSupplier`? This would probably require providing something like a `ComponentMetaDataResolver` (or maybe a `org.gradle.api.internal.artifacts.ivyservice.ivyresolve.MetadataProvider`) to the built-in rule.

bigdaz commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ivy/IvyDynamicRevisionRemoteResolveIntegrationTest.groovy:
```diff
@@ -108,6 +108,263 @@ dependencies {
+        !output.contains('Providing metadata for group:projectA:1.1')
+
+        and: "creates a new instance of rule each time"
+        !output.contains('Metadata rule call count: 2')
```
I'd like to see additional coverage here:
- Re-execute build with no changes: no network requests required, resolves to `release` version
- Publish new `integration` version and execute: only requests status file for new version and resolves to existing `release` version
- Publish new `release` version and execute: requests status file for new version and resolved to new `release version

bigdaz commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ivy/IvyDynamicRevisionRemoteResolveIntegrationTest.groovy:
```diff
@@ -108,6 +108,263 @@ dependencies {
+        then:
+
+        errorOutput.contains('Could not resolve group:projectB:latest.release')
+        failure.assertHasCause('meh: error from custom rule')
```
More failure coverage:
- Recovers from failure on subsequent execution (i.e. recovers from transient failure)
- Failure reported when requested resource does not exist

bigdaz commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ivy/IvyDynamicRevisionRemoteResolveIntegrationTest.groovy:
```diff
@@ -108,6 +108,263 @@ dependencies {
+
+        when: "resolving the same dependencies"
+        server.expectHead("/repo/status.txt", statusFile)
+        server.expectHead("/repo/status.txt", statusFile)
```
Definitely should not require multiple HEAD requests here. 
Actually, should not require _any_ network requests here, if we are honouring `cacheDynamicVersionsFor` setting.

melix replied:
I understand why this is happening. When we call `MetadataProvider#getComponentMetadata`, we avoid calling `resolve()` if possible, because of the metadata supplier. If there's one, it's called first, and we return the metadata, which is **not** a `BuildableModuleComponentMetaDataResolveResult` (because we don't need all metadata, we need a small portion of it).

As a consequence, when a supplier answers, we return directly, and perform no caching (in-memory, or persistent). This means that the supplier is always going to be called. The only thing which is cached is the remote call, and that's why we're seeing `HEAD` requests.

bigdaz replied:
As discussed on Slack, making a HEAD request to check if a resource has changed is _not_ adequate caching. We need a mechanism to avoid all network requests when a resource is cached and is up-to-date (however we define "up-to-date" for a resource).

bigdaz commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ivy/IvyDynamicRevisionRemoteResolveIntegrationTest.groovy:
```diff
@@ -108,6 +108,263 @@ dependencies {
+        server.expectHead("/repo/status.txt", statusFile)
+        checkResolve "group:projectA:1.+": "group:projectA:1.2", "group:projectB:latest.release": "group:projectB:1.1"
+
+        then: "should parse the result from cache"
```
My guess is that we're only using the result from cache because it is unchanged: this is different from the standard caching behaviour, where we avoid network requests entirely within the cache timeout period.

bigdaz commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ivy/IvyDynamicRevisionRemoteResolveIntegrationTest.groovy:
```diff
@@ -108,6 +108,263 @@ dependencies {
+        outputContains 'Providing metadata for group:projectB:2.2'
+        !output.contains('Providing metadata for group:projectB:1.1')
+        !output.contains('Providing metadata for group:projectA:1.1')
+
```
I'd like to see some test coverage for `--offline` and `--refresh-dependencies` too.

bigdaz replied:
One additional use case that is valuable to check:

1. Correctly resolve, caching resource.
2. Resolve again after the cache timeout: fails due to network error. (This happens to me when I try to execute a build when I get on a plane).
3. Resolve again with `--offline` : build succeeds

You can see some examples of this in `RecoverFromBrokenResolutionIntegrationTest`.

melix commented on subprojects/core/src/main/java/org/gradle/api/artifacts/ComponentMetadataSupplier.java:
```diff
@@ -0,0 +1,36 @@
+ * @since 3.5
+ */
+@Incubating
+public interface ComponentMetadataSupplier {
```
The detailed explanation is available in [this conversation](https://github.com/gradle/performance/issues/411#issuecomment-280315331). In short, for 2 reasons:

- the pattern is different (we don't use inline rules, but always create classes, to match the new parallel safe pattern)
- it makes it easier to find usages in IDE

bigdaz replied:
Right, but why not at least use `ComponentMetadataSupplier extends Action<? super ComponentMetadataSupplierDetails>`, so that this is a bit more consistent with our other dependency resolution rules? My assumption is that we want a common mechanism for users to supply a rule implementation: you're saying "the pattern is different" but in my mind the use-case is the same, we've just implemented a different mechanism. That's fine, but then let's do some work to move toward a common mechanism for all rule implementations.

melix commented on subprojects/core/src/main/java/org/gradle/api/artifacts/ComponentMetadataBuilder.java:
```diff
@@ -0,0 +1,47 @@
+ * @since 3.5
+ */
+@Incubating
+public interface ComponentMetadataBuilder {
```
`ComponentMetadataDetails` contains more than what we need. The `Builder` is to build metadata. The details could actually extend the builder version. The idea is to separate the concerns: a builder that allows generating the minimal set of information we need.

melix commented:
@bigdaz I added test cases, rebased onto `master` instead of `release`. Following our discussion, I did **not** introduce additional caching for resources (we'll split a different story).

bigdaz commented:
I've pushed a couple of extra commits to the branch. The main change is the implementation of proper support for `--offline`, but overriding the `ExternalResourceCachePolicy` to avoid re-checking resources when build is run offline.

I think this is OK to merge once it passes CI. One thing I'm missing is a way to provide configuration parameters to the `ComponentMetadataSupplier` instance:  the current pattern we are using for other user-supplied types is to allow the [user to provide a set of string parameters when registering the type](https://github.com/gradle/gradle/blob/324c2e974c445deb473002800ed2a0960d6e2779/subprojects/core/src/main/java/org/gradle/api/artifacts/transform/VariantTransform.java#L50-L50). We plan to improve this mechanism in the future, but I wonder if we shouldn't provide the equivalent functionality for component metadata suppliers.

bigdaz commented:
Note that I was seeing some odd flaky tests when running `:dependencyManagement:integTest` locally. It's getting late, so I didn't investigate further. Perhaps some resources are being considered up-to-date when they shouldn't be? Let's see if failures persist, and show up on CI.

melix commented:
Thanks @bigdaz the changes look great!
